algorithms/kernel/dtrees/forest/classification/oneapi/cl_kernels/df_batch_classification_kernels.cl

```diff
@@ -0,0 +1,573 @@
+            {
+                bufS[(bufIdx + sub_group_id) * nNodeProp + 0] = curFeatureId;
+                bufS[(bufIdx + sub_group_id) * nNodeProp + 1] = curFeatureValue;
+                bufS[(bufIdx + sub_group_id) * nNodeProp + 2] = (int)bestLN;
+
+                bufI[bufIdx + sub_group_id] = curImpDec;
+            }
+        }
+
+        barrier(CLK_LOCAL_MEM_FENCE);
+        if (1 < n_sub_groups && 0 == sub_group_id)
+        {
+            // first sub group for current node reduces over local buffer if required
+            algorithmFPType curImpDec = (sub_group_local_id < n_sub_groups) ? bufI[bufIdx + sub_group_local_id] : minImpDec;
+
+            int curFeatureId    = sub_group_local_id < n_sub_groups ? bufS[(bufIdx + sub_group_local_id) * nNodeProp + 0] : valNotFound;
+            int curFeatureValue = sub_group_local_id < n_sub_groups ? bufS[(bufIdx + sub_group_local_id) * nNodeProp + 1] : valNotFound;
+            int LN              = sub_group_local_id < n_sub_groups ? bufS[(bufIdx + sub_group_local_id) * nNodeProp + 2] : 0;
+
+            algorithmFPType bestImpDec = sub_group_reduce_max(curImpDec);
+
+            int bestFeatureId    = sub_group_reduce_min(bestImpDec == curImpDec ? curFeatureId : valNotFound);
+            int bestFeatureValue = sub_group_reduce_min((bestFeatureId == curFeatureId && bestImpDec == curImpDec) ? curFeatureValue : valNotFound);
+
+            bool noneSplitFoundBySubGroup = ((leafMark == bestFeatureId) && (0 == sub_group_local_id));
+            bool mySplitIsBest            = (leafMark != bestFeatureId && curFeatureId == bestFeatureId && curFeatureValue == bestFeatureValue);
+            if (noneSplitFoundBySubGroup || mySplitIsBest)
+            {
+                __global algorithmFPType * splitNodeInfo = splitInfo + nodeId * nImpProp;
+                __global algorithmFPType * nodeHistInfo  = splitInfo + nodeId * nImpProp + IMPURITY_PROPS;
+
+                splitNodeInfo[0]       = imp;
+                algorithmFPType maxVal = (algorithmFPType)0;
+                int maxInd             = 0;
+                for (int i = 0; i < nProp; i++)
+                {
+                    nodeHistInfo[i] = mrgCls[i];
+                    if (mrgCls[i] > maxVal)
+                    {
+                        maxVal = mrgCls[i];
+                        maxInd = i;
+                    }
+                }
+
+                nodeList[nodeId * nNodeProp + 2] = curFeatureId == valNotFound ? leafMark : curFeatureId;
+                nodeList[nodeId * nNodeProp + 3] = curFeatureValue == valNotFound ? leafMark : curFeatureValue;
+                nodeList[nodeId * nNodeProp + 4] = (int)LN;
+                nodeList[nodeId * nNodeProp + 5] = maxInd;
+
+                if (updateImpDecreaseRequired) nodeImpDecreaseList[nodeId] = bestImpDec;
+            }
+        }
+    }
+
+    __kernel void computePartialHistograms(const __global int * data, const __global int * treeOrder, const __global int * nodeList,
```

Comment:
We can discuss moving this functionality to service part.
I believe, we now have at least 3 places with almost the same kernels.

Reply:
agree, we can consider it for next release.